### PR TITLE
Adding return_info argument to reset to allow for optional info dict as a second return value

### PR DESCRIPTION
**Pull request by balisujohn**
This is an initial draft to make the change discussed in #1683. All changes made here are concerning the definition and implementations of the reset function in the env class in core.py.

I changed the abstract env class in core.py , so that it declares the optional argument return_info in the header

I Implemented the change for the cartpole environment.

The default behavior is to not return the info, but if the optional argument return_info is set to true, reset returns both an observation and the info dictionary.

The tests still seem to pass, and best I can tell it doesn't break subclasses of env which don't declare the optional return_info argument in the reset function header. This is an important consideration for not suddenly breaking existing third-party gym envs which subclass env.

I will try to implement this change for all the first party gym environments and extend the existing test to verify this functionality for all first party gym environments.

Any feedback is welcome.

TODO
- [x] Implement change for environments other than cartpole 
- [x] Extend test to environments other that cartpole
- [x] Make sure type checking passes
- [x] fix the merge conflict and figure out how to create a multi-case return type annotation
- [x]  Check if #2531 is going to be a problem
- [x] Ensure vectorized env compatibility.
- [x] Ensure compatibility of changes with stable-baselines3

### CODE REVIEW

**Comment by jkterry1** [comment]
@balisujohn thanks a ton for this, I asked Ariel to review it when he's back. Can you please fix the merge conflict?

**Comment by jkterry1** [comment]
@balisujohn can you please fix tests?

**Comment by jkterry1** [comment]
@balisujohn its been a week so I'm just following up on this

**Comment by jjshoots** [comment]
@jkterry1 Actually I think the only test that's failing is black, which isn't too hard to do if you'd like to merge it soon.

**Comment by balisujohn** [comment]
This is not ready to merge. On my local branch I extended the change to a lot of other first party environments and it broke a lot of vectorized environment tests. Changes are necessary to the vectorized environment code to make the existing tests pass, and new tests will need to be added to make sure the vectorized environments also respond appropriately to the return_info argument in the reset function.

File: gym/core.py
Comment (RedTachyon) [inline gym/core.py:33]: Typo

File: gym/core.py
Comment (RedTachyon) [inline gym/core.py:73]: It shouldn't be Optional[bool], just bool (but still with a default of False)

Despite the appearances, `Optional` doesn't mean that the argument is optional, but that it can also take the value `None`, which doesn't make sense here

File: gym/envs/classic_control/cartpole.py
Comment (RedTachyon) [inline gym/envs/classic_control/cartpole.py:162]: So this will be the problematic part, since we'll need to patch this in all environments (not a big deal), but also make sure that third-party environments won't have problems with it (and I don't think there's anything we can do to automatically make them obey the new contract of `reset`)

**Review by RedTachyon — COMMENTED** [review]
I'm not sure what the issues with vector API are, but I can very easily believe there are many. Overall I'm not fully convinced it's good to introduce this in a 0.x release, since it is a major change of the core behavior, even if it's optional

File: gym/core.py
Comment (balisujohn) [reply gym/core.py:73]: Good catch, I will fix this.

File: gym/envs/classic_control/cartpole.py
Comment (balisujohn) [reply gym/envs/classic_control/cartpole.py:162]: I've started work on extending this patch to all first-party environments, so far, I haven't found any which return non-empty metadata, which is interesting.

File: gym/core.py
Comment (balisujohn) [reply gym/core.py:33]: good catch, will fix

**Comment by jkterry1** [comment]
@balisujohn Hey man, I'm just following up on the status of this. Relatedly, tests still aren't passing

**Comment by balisujohn** [comment]
@jkterry1 I'll see if I can get what I have pushed by the end of the week. The sticking point is the AsyncVectorEnv, it's proving a bit hard to get the worker threads to also return a the metadata dict, but I expect I can fix it.

**Comment by jkterry1** [comment]
@balisujohn could you please push whatever you have in the fairly near future? We need to include this in the next release of Gym, so I'm going to have Ariel finish whatever you haven't sometime next week

**Comment by jkterry1** [comment]
(and could you please fix the new merge conflicts too?)

**Comment by balisujohn** [comment]
So basically the return_info argument is implemented for all first party environments, and all implemented first party environments aside from the unit-test environments are tested in the test test_reset_info in test_envs.py. 

There is some work towards extending this functionality to both wrappers and vector environments, though tests still need to be written to check if the wrappers and vector environments correctly return metadata from the reset function. 

async_vector_env.py proved particularly difficult to instrument to optionally return metadata on reset, as the metadata must be passed back from worker threads conditionally, based on the the value of the return_info argument being sent to worker threads, so for now it is left unimplemented.

- [ ] return_info = True test for sync_vector_env.py
- [ ] return_info = True test for async_vector_env.py
- [ ] return_info = True implementation for async_vector_env.py
- [ ] return_info = True test for normalize.py 
- [ ] decide if record_episode_statistics should support return_info argument (should we return the initial info from calling reset)

**Comment by balisujohn** [comment]
Also of note the only first party environments that return non-empty metadata dicts under any circumstance as far as I could tell were a subset of the toytext environments(frozen lake for example). It might be worth reviewing my decision to return a probability 1 as the metadata value returned from the reset function. It seems to be logically consistent with the meaning of the "prob" metadata parameter, but it's worth double checking.

**Comment by balisujohn** [comment]
And a final note: it is of concern whether or not the wrappers/vector envs assume that any particular optional arguments are available the reset functions of third party environments they are wrapping. It seems like the sync_vector_env assumes that  the 'seed' keyword arguments will be available, which is a breaking change for third party environments which predate the creation of the 'seed' keyword argument in reset. So it's important to note I believe we have already made breaking changes to at least sync_vector_env.

**Comment by RedTachyon** [comment]
@balisujohn I can't test it right now, but when I was editing the code for vector envs, the intention was to keep it backwards compatible, and when I'm reviewing it now, I'm like 80% that it is. If someone's using an old style environment that doesn't support these arguments, as long as they don't try to explicitly pass a seed, it should all be fine. 

To make old environments compatible with the changed API functionality, they need to be updated, perhaps we'll also make a wrapper to do it automatically in gym.make.

**Comment by balisujohn** [comment]
@RedTachyon I'll see if I can create a minimal example that either confirms or disproves my claim. The main things left for this pull request other than a general review for quality are to decide if/how to implement the optional return_info behavior for atari_preprocessing.py and also to implement and test the return_info behavior in async_vec_env.py. I'm curious what your opinion is regarding the atari_preprocessing.py wrapper(like where are the atari envs implemented and do we have the ability to make them conform to the new standards?).

**Comment by RedTachyon** [comment]
@balisujohn So I did a test with pybullet envs, and it works the way I expected. A code snippet like this (with the new version of gym) works:

```python
import pybullet_envs
import gym
env = gym.vector.make("HopperBulletEnv-v0")
env.reset()
```

but calling
```python
env.reset(seed=0)
```

causes a crash. So old-style environments still work with old-style resets. And if you want to do a new-style reset with an old-style behavior, that requires some more thought, but isn't strictly speaking necessary for backwards compatibility. 


As for atari - I don't use it and can't really test it right now, but the maintainer of ALE is in the loop and will update it when the time comes. For the moment it should work the same way the pybullet example does.

**Comment by balisujohn** [comment]
I think this is ready for final review. I checked and at least for cartpole stable-baselines3 works fine. It seems like my fears about compatibility with out of data environments being broken by sync_vector_env were unfounded. Barring any mistakes, there are implementations and tests of the optional return_info behavior for all first-party environments, wrappers, and vector environments, except for atari_preprocessing.py. I think we should get feedback from someone who's more focused on the atari side of things, because it seems like the existing tests for atari_preprocessing.py have been disabled since the folder atari in the envs folder was deleted in a previous commit.

**Comment by balisujohn** [comment]
It seems like the testing environment doesn't have mujoco-py installed, meaning the tests fail when testing the reset behavior on the mujoco environments. Shall I remove the tests for the first part mujoco environments, or would it be preferred to add mujoco-py as a testing environment dependency?

**Comment by jkterry1** [comment]
@balisujohn all tests are currently disabled for mujoco environments AFAIK

**Comment by balisujohn** [comment]
I added a test that tests the reset behavior for all first party environments including the mujoco environments, so I can remove those environments from the test.

**Comment by jkterry1** [comment]
Right, please pull the mujoco tests. They'll be removed from Gym in some number of months anyways

**Comment by jkterry1** [comment]
@balisujohn tests failed due to linting

**Comment by balisujohn** [comment]
It seems like my version of black from python3.9 disagrees about how to format  gym/utils/atomic_write.py
 with the version being used in the pre-commit deployment step. (black version 21.12b0)

**Comment by jkterry1** [comment]
@balisujohn I think that file is dead code, right? Why don't you just remove it.

**Comment by balisujohn** [comment]
The linting should be fixed; I'm not sure what atomic_write.py does, this was the first time I noticed it. Maybe removing it would be a good first issue, to get new contributors familiar with the pull request process?

**Comment by balisujohn** [comment]
@jkterry1  do you want this functionality added for atari_preprocessing.py? Otherwise, I think this pull request is ready for final review.

**Comment by jkterry1** [comment]
@balisujohn let's have Jesse do that after this is merged

File: gym/core.py
Comment (XuehaiPan) [inline gym/core.py:77]:
```suggestion
        return_info: bool = False,
```

`Optional[bool]` means `Union[bool, None]`. Same changes to the remaining type hints.

File: tests/envs/test_envs.py
Comment (RedTachyon) [inline tests/envs/test_envs.py:60]: I'm pretty sure there's a pattern somewhere in tests to automatically run tests on all valid envs, I think it's better to use that. This would need to be updated whenever one of these envs needs a version bump.

File: tests/wrappers/test_record_episode_statistics.py
Comment (RedTachyon) [inline tests/wrappers/test_record_episode_statistics.py:33]: So two points, applying to this specific line and the ones similar in other parts of the tests.
1. As mentioned in another comment, you can parametrize the test so that it runs for all environments, so that they're all checked and not just CartPole
2. I think a better to check for validity of the observation would be `env.observation_space.contains(obs)`. It generalizes to all other weirder action spaces (for example this version would fail with a Dict action space)

File: tests/wrappers/test_record_video.py
Comment (RedTachyon) [inline tests/wrappers/test_record_video.py:36]: I might be missing something, does this line actually do anything? Directly accessing a `_parameter` is a bit iffy

**Review by RedTachyon — CHANGES_REQUESTED** [review]
Pretty good job overall, but there's a few changes that need to be done. Particularly important I think is making the tests run for all envs, since this can be an easy-to-break thing when we change/add envs.

**Comment by balisujohn** [comment]
Thanks for the code review! I'll try to make the requested fixes soon.

**Comment by jkterry1** [comment]
@balisujohn there's also merge conflicts

**Comment by jkterry1** [comment]
@balisujohn Do you know how much longer making the fixes here will take? This is something that we need for the next upcoming release.

File: gym/core.py
Comment (balisujohn) [reply gym/core.py:77]: Sounds good, it looks like I missed that one; will fix.

File: gym/core.py
Comment (balisujohn) [reply gym/core.py:79]: Fixing this.

**Comment by balisujohn** [comment]
I updated the code based on the recommendations made in the code review. I merged the code with the current version in master. All tests pass when 

````
images = env.call("render", mode="rgb_array")
````

and the corresponding asserts in test_call_async_vector_env are disabled. When enabled, test_call_async_vector_env passes when only the tests in test_async_vector_env are called but fails with the following output:

````
tests/vector/test_async_vector_env.py ........XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":1"
      after 766 requests (766 known processed) with 1 events remaining.
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":1"
      after 766 requests (766 known processed) with 4 events remaining.
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":1"
      after 767 requests (767 known processed) with 6 events remaining.
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":1"
      after 766 requests (766 known processed) with 7 events remaining.
FXIO:  fatal IO error 0 (Success) on X server ":1"
      after 766 requests (766 known processed) with 0 events remaining.
````

when being run with the entire test battery. This seems like an intermittent error related to rendering(potentially there is some complication making the render function fail in the vectorized asynchronous setting). Is it apparent to anyone why this is happening?

File: tests/wrappers/test_record_video.py
Comment (balisujohn) [reply tests/wrappers/test_record_video.py]: This was just due to a naive copy of another test being used as a template.  It is now removed.

File: gym/vector/sync_vector_env.py
Comment (balisujohn) [reply gym/vector/sync_vector_env.py]: You are totally right, It should now be fixed.

File: tests/envs/test_envs.py
Comment (balisujohn) [reply tests/envs/test_envs.py]: removed.

File: tests/envs/test_envs.py
Comment (balisujohn) [reply tests/envs/test_envs.py]: Good catch, I made this change.

File: tests/wrappers/test_record_episode_statistics.py
Comment (balisujohn) [reply tests/wrappers/test_record_episode_statistics.py]: 1. I switched the direct for all first party environments in test_envs to the parameterization decorator. For now I'm leaving the wrapper tests as single-environment tests, since many existing tests are like this, and it might make the tests take much longer. If people think it's worthwhile, I'm happy to add more coverage such that we try every first party env with every wrapper/vector env.

2. I made this switch for all the direct first party env tests  wrapper tests except normalize.py because it seems to not contain the observation in the observation space! This might be an error worth investigating. I also didn't implement this for the vector env tests.

**Comment by RedTachyon** [comment]
I'm slightly confused, I don't see any rendering-related changes in the code here. Is the error only local on your machine, or did it also ever happen in the github pipeline?

If it's the former, I'd say if the error doesn't happen in the pipeline here, then it doesn't exist. I regularly get some errors on my local setup, but figured it's just a matter of my configuration not being good for some of the more esoteric features (shared memory, displays and all that nonsense)


Overall, if it's just a local error, my comments seem to be addressed so overall should be good to go.

**Comment by balisujohn** [comment]
It's definitely local, because I was able to reproduce it with the version of gym currently in OpenAI gym master. Must be something weird with my Xorg installation.

**Pull request merged**
